streams/src/main/java/org/apache/kafka/streams/KeyValue.java

```diff
@@ -42,4 +42,21 @@ public KeyValue(K key, V value) {
     public String toString() {
         return "KeyValue(" + key + ", " + value + ")";
     }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof KeyValue) {
+            KeyValue otherKV = (KeyValue) other;
+
+            return key.equals(otherKV.key) && value.equals(otherKV.value);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        long n = ((long) key.hashCode() << 32) | (long) value.hashCode();
```

Comment:
Same here: NPEs are possible.

Assuming that we do allow null values for `key` and `value`, the tradition says to assume `0` for the hashcode of the null field (see Effective Java, Item 9, p. 47).

One way to achieve that is by using [Objects.hashCode(Object o)](http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hashCode%28java.lang.Object%29), which returns `0` when `o == null`.

So we could update the code to:

```java
long n = ((long) Objects.hashCode(key) << 32) | (long) Objects.hashCode(value);
```

That said, we should perhaps use the helper function [`Objects.hash()`](http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash%28java.lang.Object...%29) directly, which was introduced in 1.7 (I'd vote for this latter implementation due to simplicity and reusing Java's internal functionality for this):

```java
@Override
public int hashCode() {
    return Objects.hash(key, value);
}
```